pkg/lexer/state.go

```diff
@@ -0,0 +1,326 @@
+// Copyright © 2021 Rak Laptudirm <raklaptudirm@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lexer
+
+import (
+	"fmt"
+	"unicode"
+
+	"github.com/raklaptudirm/mash/pkg/token"
+)
+
+var ErrUnexpectedEOF = fmt.Errorf("unexpected EOF")
+
+func (l *lexer) run() {
+	lexBase(l)
+	close(l.Tokens)
+}
+
+func lexBase(l *lexer) {
+next:
+	r := l.peek()
+	switch {
+	case r == eof:
+		l.emit(token.SEMICOLON)
+		l.emit(token.EOF)
+		return
+	case unicode.IsSpace(r):
+		consumeAllSpace(l)
+	case r == '#':
+		lexComment(l)
+	case isAlphabet(r):
+		consumeWord(l)
+
+		word := l.literal()
+		if token.IsKeyword(word) {
+			l.emit(token.Lookup(word))
+			lexStmt(l)
+			break
+		}
+
+		l.emit(cmdOpLookup(word))
+		fallthrough
+	default:
+		lexCmd(l)
+	}
+
+	goto next
+}
+
+func isAlphabet(r rune) bool {
+	return r > 'A' && r < 'Z' || r > 'a' && r < 'z'
+}
+
+func lexStmt(l *lexer) {
```

Comment:
I don't see this implementation as too different from what I did before... The only difference I see is that my implementation was calling the methods/functions somewhere else whereas you are calling the function here. I don't understand why you didn't like the previous approach then?

Reply:
Actually, after that conversation, I realized that implementing the parser for mash using the `Next()` lexing approach will be problematic because of the "lex statements and commands differently" nuance. So I initially switched to the state machine approach that you used till e8bf4adf920c0b865ee04b0e2281989bc6f0ec4f. After that, I found that calling the functions explicitly was more extendable and made some stuff easier, so I changed the implementation to the current one.